### PR TITLE
Streaming JSON serializers

### PR DESCRIPTION
**Pull request by bugaevc**
Two new classes, `JsonArraySerializer` and `JsonObjectSerializer`, that can be used to serialize JSON to a builder without generating the whole structure in memory in advance. You can use them like this:

```c++
SomeBuilder builder;
JsonArraySerializer array { builder };
array.add(35);
array.add("awesome");

// A nested array.
{
    JsonArraySerializer nested_array = array.add_array();
    nested_array.add(42);
}

// A nested object.
{
    JsonObjectSerializer nested_object = array.add_object();
    nested_object.add("foo", "bar");
}
```

Note how instead of first creating the nested array/object (all in memory!) and then doing `array.add(move(nested))` we get the nested object from the parent. Both serializer classes are just a few bytes in size, so doing this is really cheap, and `add()` writes content directly to the builder.

This should help with https://github.com/SerenityOS/serenity/issues/484

`JsonObject::serialize()` and `JsonArray::serialize()` themselves now wrap these serializer classes. It's not more (or less) efficient to do this if you already have the array/object in memory, but this way we avoid duplicating the logic.

### CODE REVIEW

File: AK/JsonArraySerializer.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            finish();
+    }
+
+    void add(const JsonValue& value)
+    {
+        begin_item();
+        value.serialize(m_builder);
+    }
+
+    JsonArraySerializer<Builder> add_array()
```
Comment (awesomekling) [inline AK/JsonArraySerializer.h]: Let's add `[[nodiscard]]` this so the compiler warns you if you call it without using the return value.

File: AK/JsonArraySerializer.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        begin_item();
+        return JsonArraySerializer(m_builder);
+    }
+
+    // Implemented in JsonObjectSerializer.h
+    JsonObjectSerializer<Builder> add_object();
```
Comment (awesomekling) [inline AK/JsonArraySerializer.h]: `[[nodiscard]]`

File: AK/JsonArraySerializer.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace AK {
+
+template<typename Builder>
+class JsonObjectSerializer;
+
+template<typename Builder>
+class JsonArraySerializer {
+public:
+    JsonArraySerializer(Builder& builder)
```
Comment (awesomekling) [inline AK/JsonArraySerializer.h]: This should be marked `explicit` to prevent accidental conversion.

File: AK/JsonObjectSerializer.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+#include <AK/JsonArraySerializer.h>
+#include <AK/JsonValue.h>
+
+namespace AK {
+
+template<typename Builder>
+class JsonObjectSerializer {
+public:
+    JsonObjectSerializer(Builder& builder)
```
Comment (awesomekling) [inline AK/JsonObjectSerializer.h]: This should be marked `explicit` to prevent accidental conversion.

File: AK/JsonObjectSerializer.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            finish();
+    }
+
+    void add(const StringView& key, const JsonValue& value)
+    {
+        begin_item(key);
+        value.serialize(m_builder);
+    }
+
+    JsonArraySerializer<Builder> add_array(const StringView& key)
```
Comment (awesomekling) [inline AK/JsonObjectSerializer.h]: `[[nodiscard]]`

File: AK/JsonObjectSerializer.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        begin_item(key);
+        return JsonArraySerializer(m_builder);
+    }
+
+    JsonObjectSerializer<Builder> add_object(const StringView& key)
```
Comment (awesomekling) [inline AK/JsonObjectSerializer.h]: `[[nodiscard]]`

File: AK/JsonArraySerializer.h (hunk 1 above)
Comment (bugaevc) [reply AK/JsonArraySerializer.h]: Why, maybe you want to add an empty array

Comment (awesomekling) [reply AK/JsonArraySerializer.h]: Hmm.. You know what.. I guess maybe you would want to do that. Never mind about `[[nodiscard]]`. :)

**Comment by awesomekling** [comment]
We should improve this further by not requiring everyone to go via `JsonObjectSerializer::add(StringView, JsonValue)` since that function forces implicit construction of a `JsonValue` object when you call e.g `add("foo", some_boolean)`

**Comment by awesomekling** [comment]
That's a bad example actually, bool conversion is fine, the problem is with strings.
`add("foo", "bar")` will construct a temporary `JsonValue` to contain `"bar"` which means we allocate a `String` object and throw it away immediately after serializing. Sad!

**Comment by bugaevc** [comment]
That was my original plan, but seeing how much boilerplate it was, and that `JsonValue` doesn't allocate and is generally is lightweight, I decided to do it this way. You're definitely welcome to revisit this decision though :)

> That's a bad example actually, bool conversion is fine, the problem is with strings.
add("foo", "bar") will construct a temporary JsonValue to contain "bar" which means we allocate a String object and throw it away immediately after serializing. Sad!

Hmm, yes you're right about strings.

**Pull request merged**